### PR TITLE
`finalize` callbacks.

### PR DESCRIPTION
**Pull request by oscardssmith**
fixes https://github.com/SciML/DiffEqBase.jl/issues/931 (although we probably need a similar line added for `Sundials.jl`)

### CODE REVIEW

File: test/integrators/discrete_callback_dual_test.jl [hunk 1]
```diff
@@ -7,11 +7,11 @@ using OrdinaryDiffEq, Test, ForwardDiff
 u0 = 1.0
 tspan = (0.0, 1.0)
 p = 1.0
-
+X = 0
```
Comment (staticfloat) [inline test/integrators/discrete_callback_dual_test.jl:10]:
```suggestion
times_finalize_called = 0
```

**Comment by ChrisRackauckas** [comment]
Yes Sundials and StochasticDiffEq

**Comment by staticfloat** [comment]
Does this also close https://github.com/SciML/OrdinaryDiffEq.jl/issues/1941 ?

**Comment by ChrisRackauckas** [comment]
yes

**Pull request merged**